gce/survival-training/README.md

```diff
@@ -0,0 +1,93 @@
+
+There are five components to our resilient training jobs:
+
+1. [GCE custom images](https://cloud.google.com/compute/docs/images#custom_images)
+
+1. A trainer command line interface
+
+1. [GCE startup scripts](https://cloud.google.com/compute/docs/startupscript)
+
+1. [GCE instance metadata](https://cloud.google.com/compute/docs/storing-retrieving-metadata)
+
+1. [Cloud Source Repositories](https://cloud.google.com/source-repositories/)
+
+
+As a part of this process, if you have not already done so, you will be prompted to enable both the Compute Engine API as well as the Cloud Source Repositories API for your Google Cloud Platform project.
+
+We will be working through the [Google Cloud SDK](https://cloud.google.com/sdk/). If you have not already done so, it is worth installing.
+
+
+### Custom images
+
+Your training job may require specific versions of libraries, like [tensorflow](https://www.tensorflow.org/) or [scikit-learn](http://scikit-learn.org). Additionally, in order to take advantage of instances with attached GPUs, you will have to have hardware-specific drivers installed on the instance. [Custom images](https://cloud.google.com/compute/docs/images#custom_images) are a convenient solution to this problem of environment reproduction within a GCE virtual machine.
+
+If you do have requirements (in addition to simply the OS) in your training environments, it is highly recommended that you create a VM image that you can use to spawn new instances without having to go through the tedious configuration process every time. The examples following this strategy section will demonstrate the creation of such images so, if this seems daunting, don't worry.
+
+
+### Trainer CLI
+
+Our strategy will have the GCE instance automatically kick off your training job on startup. As part of this kick off, it will have to provide the job with the appropriate parameters -- how many steps to train for, how often to checkpoint the model, which data to train the model with, what hyperparameters should be used in training the model, etc.
+
+It is helpful, if you intend to do this more than once, to assert a common semantical system across all your trainers. We provide a [boilerplate CLI](./dummy/train.py) in this repo. This interface is very similar to that provided by [TensorFlow estimators](https://www.tensorflow.org/programmers_guide/estimators), but can be used even with other frameworks. In the sections that follow, we will provide you with both TensorFlow and with scikit-learn examples that demonstrate its use.
+
+
+### Cloud Source Repositories
+
+We need a mechanism by which we can get our trainer code into the GCE instance. There are several solutions available to us:
+
+1. We could simply store the trainer application on the custom image from which we create the GCE instance.
+
+1. We could store the code in a [GCS](https://cloud.google.com/storage/) bucket and have the startup script download it to the instance before doing anything else.
```

Comment:
A response to issues both you and @dizcology have raised around the complexity of source repos:

I think the source repo also fits into this decoupling picture that you mentioned in your comment @puneith  - it is only responsible for the versioning. It serves exactly the same role as the GCS bucket in your comment, except that it is perhaps less familiar to users than GCS (which *is* a serious shortcoming).

In general, if someone wants to use the code here to wrap up existing models, they *are* going to have to do some work to wrap it up in the dummy trainer, and it will be of substantial value to them to have version control available to them as part of that process. I also hesitate to leave it to them completely when it comes to figuring out versioning for their deployed models. For these reasons, I think source repos also have advantages sufficient to overcome the shortcoming of unfamiliarity.

Still, it makes sense to offer a second option canonically of deploying from GCS instead. I suggest adding that as a feature in a separate PR once we are done with this one. It shouldn't block this PR. I think I can do that this week.

Reply:
The issue is not just about unfamiliarity, versioning systems in general are never used (and there is always a decoupling layer of build, test, release systems) to deploy software directly. And so the tight coupling pattern like I said doesn't give people any room to deploy software however they want. Basically when you assume trainer package somehow got to GCS then you would have introduced decoupling and not made any assumptions about how the package got to GCS (Build, Test, Release). There is also a reason why SemVer is a thing because version control systems never provide the versioning which is needed for software.

Reply:
Assuming that it somehow gets into GCS is equivalent to assuming that it somehow ends up in a source repo, though.

My intention was not to force anyone to use source repos, but rather suggest them as the best possible solution, and I do mention other possibilities in the doc.

To the extent that this introduces coupling, so does the GCS method. For example, what I will have to do in order to get a trainer from gcs is a `gsutil -m cp -r ...` instead of a `gcloud source repos clone`. I'm not seeing how one is more or less coupled than the other.

Also Tensorport uses GitHub repos and commit hashes/tags to specify which version of a model gets used in a training job, so I don't think it's as cut and dried as versioning systems never being used for such deployments.

In general, one has to give a lot of thought to versioning (of different kinds - training data, eval data, model parameters, model architecture) in machine learning workflows - e.g. [hackernews thread](https://news.ycombinator.com/item?id=13821217).

Reply:
>>Assuming that it somehow gets into GCS is equivalent to assuming that it somehow ends up in a source repo, though.
This assumption is actually not true. There is an entire workflow (see Gitflow and Github flow) dedicated to code development on Git and an entire build, test, release, deploy cycle to ultimately produce an artifact on some storage (eg., Disk, HDFS, GCS, S3 etc) which can be consumed by downstream systems. Its true that versioning is not straightforward but its not an unsolvable problem and therefore e.g., SemVer. Its software development at the end of the day and canonical workflows apply directly. Anyways, I don't want to rathole and digress too much from the main goal here. But I wanted to highlight this so that we know what these choices mean for our users.

Reply:
Although this conversation is a digression, I think it is a useful one. I have heard the ML Engine team talk about source repos, as well, in a related capacity. We should have some kind of a stance on this.

I'm okay either way, because it's a pretty minimal change to switch over to using GCS or GitHub or whatever in this code. I just think that being able to tag commits for use in training jobs is extremely valuable. That way you don't have to go through contortions to remember things like "did I use dropout between the third and fourth layers of my model in that training job?" This particular example doesn't show that, but it makes it possible directly as opposed to indirectly (use Gitflow being more prescriptive).

Also, the problem is not that versioning is unsolvable but rather that there are so many possible solutions. It is not clear that one is better than another (or that one has won out).